### PR TITLE
Add public api to register global listener

### PR DESCRIPTION
**Pull request by shalehaha**
Add an API to MobileCore to allow registering of an event listener. Mainly to support the scenario where an extension need to support a public permanent callback.

### CODE REVIEW

File: AEPCore/Sources/core/MobileCore.swift [hunk 1]
```diff
@@ -94,13 +94,23 @@ public final class MobileCore: NSObject {
     ///   - responseCallback: Callback to be invoked with `event`'s response `Event`
     @objc(dispatch:responseCallback:)
     public static func dispatch(event: Event, responseCallback: @escaping (Event?) -> Void) {
-        EventHub.shared.registerResponseListener(triggerEvent: event, timeout: 1) { event in
+        EventHub.shared.registerResponseListener(triggerEvent: event, timeout: 2) { event in
```
Comment (sbenedicadb) [inline AEPCore/Sources/core/MobileCore.swift:97]: should we make this timeout configurable? or at least explain that it exists in the docs for the method?

Comment (shalehaha) [reply AEPCore/Sources/core/MobileCore.swift]: will address this in a separate pr

File: AEPCore/Sources/eventhub/AEPError.swift [hunk 2]
```diff
@@ -20,5 +20,9 @@ import Foundation
     case callbackTimeout = 1
     case callbackNil = 2
     case none = 3
+    case serverError = 4
```
Comment (pfransenadb) [inline AEPCore/Sources/eventhub/AEPError.swift:23]: Are these an artifact from some other work?  I don't see them used anywhere in this PR.

File: AEPCore/Sources/eventhub/EventHub.swift [hunk 3]
```diff
@@ -148,6 +148,23 @@ final class EventHub {
         responseEventListeners.append(responseListenerContainer!)
     }
 
+    /// Registers an `EventListener` which will be invoked whenever a event with matched type and source is dispatched
+    /// - Parameters:
+    ///   - type: An `String` indicates the event type the current listener is listening for
+    ///   - source: An `String` indicates the event source the current listener is listening for
+    ///   - listener: An `EventResponseListener` which will be invoked whenever the `EventHub` receives a event with matched typd and source
+    func registerEventListener(type: String, source: String, listener: @escaping EventListener) {
+
```
Comment (pfransenadb) [inline AEPCore/Sources/eventhub/EventHub.swift:157]: nit: remove blank line.

File: AEPCore/Sources/eventhub/EventSource.swift [hunk 4]
```diff
@@ -22,6 +22,7 @@ public class EventSource: NSObject {
     public static let requestIdentity = "com.adobe.eventSource.requestIdentity"
     public static let requestProfile = "com.adobe.eventSource.requestProfile"
     public static let requestReset = "com.adobe.eventSource.requestReset"
+    public static let notification = "com.adobe.eventSource.notification"
```
Comment (pfransenadb) [inline AEPCore/Sources/eventhub/EventSource.swift:25]: Unrelated to this PR?

File: AEPCore/Sources/eventhub/EventType.swift [hunk 5]
```diff
@@ -34,6 +34,7 @@ public class EventType: NSObject {
     public static let userProfile = "com.adobe.eventType.userProfile"
     public static let places = "com.adobe.eventType.places"
     public static let offerDecisioning = "com.adobe.eventType.offerDecisioning"
+    public static let edge = "com.adobe.eventType.edge"
```
Comment (pfransenadb) [inline AEPCore/Sources/eventhub/EventType.swift:37]: Unrelated?

**Review by pfransenadb — CHANGES_REQUESTED** [review]
Minor nit + clarity of PR.

**Comment by shalehaha** [comment]
@pfransenadb fixed. removed the code that is not related to this pr.

File: AEPCore/Sources/core/MobileCore.swift [hunk 6]
```diff
@@ -101,6 +101,16 @@ public final class MobileCore: NSObject {
         EventHub.shared.dispatch(event: event)
     }
 
+    /// Registers an `EventListener` which will be invoked whenever a event with matched type and source is dispatched
+    /// - Parameters:
+    ///   - type: An `String` indicates the event type the current listener is listening for
```
Comment (sbenedicadb) [inline AEPCore/Sources/core/MobileCore.swift:106]: `An String indicates` > `A String indicating`

File: AEPCore/Sources/core/MobileCore.swift [hunk 7]
```diff
@@ -101,6 +101,16 @@ public final class MobileCore: NSObject {
...
+    ///   - source: An `String` indicates the event source the current listener is listening for
```
Comment (sbenedicadb) [inline AEPCore/Sources/core/MobileCore.swift:107]: same here

File: AEPCore/Sources/core/MobileCore.swift [hunk 8]
```diff
@@ -101,6 +101,16 @@ public final class MobileCore: NSObject {
...
+    ///   - listener: An `EventResponseListener` which will be invoked whenever the `EventHub` receives a event with matched typd and source
```
Comment (sbenedicadb) [inline AEPCore/Sources/core/MobileCore.swift:108]: `typd` > `type`

File: AEPCore/Sources/eventhub/EventHub.swift [hunk 9]
```diff
@@ -148,6 +148,22 @@ final class EventHub {
         responseEventListeners.append(responseListenerContainer!)
     }
 
+    /// Registers an `EventListener` which will be invoked whenever a event with matched type and source is dispatched
+    /// - Parameters:
+    ///   - type: An `String` indicates the event type the current listener is listening for
```
Comment (sbenedicadb) [inline AEPCore/Sources/eventhub/EventHub.swift:153]: same fixes needed for copy/pasted docs

File: AEPCore/Sources/eventhub/EventHub.swift [hunk 10]
```diff
@@ -148,6 +148,22 @@ final class EventHub {
...
+    ///   - source: An `String` indicates the event source the current listener is listening for
+    ///   - listener: An `EventResponseListener` which will be invoked whenever the `EventHub` receives a event with matched typd and source
+    func registerEventListener(type: String, source: String, listener: @escaping EventListener) {
+        eventHubQueue.async {
+            // use the event hub placeholder extension to hold all the listeners register from the public API
```
Comment (sbenedicadb) [inline AEPCore/Sources/eventhub/EventHub.swift:158]: `register` > `registered`

**Review by sbenedicadb — APPROVED** [review]
looks good, just a few updates to the comments would be nice

File: AEPCore/Sources/core/MobileCore.swift [hunk 11]
```diff
@@ -101,6 +101,16 @@ public final class MobileCore: NSObject {
         EventHub.shared.dispatch(event: event)
     }
 
+    /// Registers an `EventListener` which will be invoked whenever a event with matched type and source is dispatched
+    /// - Parameters:
+    ///   - type: An `String` indicating the event type the current listener is listening for
```
Comment (sbenedicadb) [inline AEPCore/Sources/core/MobileCore.swift:106]: `An` should be `A`
